### PR TITLE
Solution

### PR DESCRIPTION
**Pull request by dissom**

### CODE REVIEW

**Review by vasylhnatiuk — CHANGES_REQUESTED** [review]
fix flake8 tests

File: app/car.py [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+    fuel_consumption: float
+
+    def trip_cost(
+            self,
+            start_point: list[int],
+            end_point: list[int],
+            fuel_price: float
+    ) -> float:
+        distance = sqrt((start_point[0] - end_point[0]) ** 2
+                        + (start_point[1] - end_point[1]) ** 2)
```
Comment (vsmutok) [inline app/car.py]:
```suggestion
        distance = dist(start_point, end_point)
```

File: app/customer.py [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+            print()
+            self.make_purchase(cheapest_shop)
+            date = (datetime(
+                2021, 4, 1, 12, 33, 41
+            ).strftime("%m/%d/%Y %H:%M:%S"))
+            print(f"Date: "
+                  f"{date}"
+                  )
+            print(f"Thanks, {self.name}, for your purchase!")
+            print("You have bought:")
```
Comment (vsmutok) [inline app/customer.py]: combine into one `print()`

File: app/customer.py [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+            print(f"{self.name}'s trip to the "
+                  f"{shop.name} costs {round(total_cost, 2)}")
+
+            if total_cost <= min_cost and total_cost <= self.money:
+                min_cost = total_cost
+                cheapest_shop = shop
+
+        if cheapest_shop:
+            print(f"{self.name} rides to {cheapest_shop.name}")
+            print()
```
Comment (vsmutok) [inline app/customer.py]: combine into one `print()`

File: app/customer.py [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+                total_price = (
+                    int(total_price) if total_price == int(total_price)
+                    else float(total_price)
+                )
+                print(f"{self.product_cart[product]} "
+                      f"{product}s for {total_price} dollars")
+            print(f"Total cost is {self.make_purchase(cheapest_shop)} dollars")
+            print("See you again!")
+            print()
+            print(f"{self.name} rides home")
```
Comment (vsmutok) [inline app/customer.py]: combine into one `print()`

File: app/customer.py [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+            self.money -= min_cost
+            print(f"{self.name} now has {round(self.money, 2)} dollars")
+            print()
```
Comment (vsmutok) [inline app/customer.py]: combine into one `print()`

File: app/main.py [hunk 6]
```diff
@@ -1,3 +1,44 @@
...
+        ) for customer_data in customers_data
+    ]
+
+    shops_data = config.get("shops")
+
+    shops = [Shop(
+        name=shop.get("name"),
+        location=shop.get("location"),
+        products=shop.get("products")
+    ) for shop in shops_data]
```
Comment (vsmutok) [inline app/main.py]:
```suggestion
    shops = [Shop(**shop) for shop in shops_data]
```

File: app/main.py [hunk 7]
```diff
@@ -1,3 +1,44 @@
...
+        Customer(
+            name=customer_data.get("name"),
+            product_cart=customer_data.get("product_cart"),
+            location=customer_data.get("location"),
+            money=customer_data.get("money"),
+            car=Car(
+                brand=customer_data.get("car").get("brand"),
+                fuel_consumption=customer_data.get("car").get
+                ("fuel_consumption")
+            )
```
Comment (vsmutok) [inline app/main.py]:
```suggestion
            car=Car(**customer_data.get("car"))
```

File: app/main.py [hunk 8]
```diff
@@ -1,3 +1,44 @@
...
+from app.shop import Shop
+from app.customer import Customer
+
+
+def shop_trip() -> None:
+    try:
+        with open("app/config.json", "r") as file:
+            config = json.load(file)
+    except FileNotFoundError:
+        print("No Such File!")
```
Comment (vsmutok) [inline app/main.py]: better to raise Exception, to abort script execution

File: app/customer.py [hunk 9]
```diff
@@ -0,0 +1,80 @@
+from __future__ import annotations
```
Comment (vsmutok) [inline app/customer.py]: remove redundant import

**Review by vsmutok — CHANGES_REQUESTED** [review]

File: app/customer.py [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+    money: int
+    car: Car
+
+    def make_purchase(self, shop: Shop) -> int | float:
+        total_purchase = 0
+        for product, price in shop.products.items():
+            if product in self.product_cart:
+                product_cost = price * self.product_cart[product]
+                total_purchase += product_cost
+        return total_purchase
```
Comment (vsmutok) [inline app/customer.py]:
```suggestion
        return sum(
            price * self.product_cart[product]
            for product, price in shop.products.items()
            if product in self.product_cart
        )
```

File: app/customer.py [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+
+        for shop in shops:
+            trip_to_shop = self.car.trip_cost(
+                self.location,
+                shop.location,
+                fuel_price
+            )
+            total_cost = round(trip_to_shop * 2, 2) + self.make_purchase(shop)
+            print(f"{self.name}'s trip to the "
+                  f"{shop.name} costs {round(total_cost, 2)}")
```
Comment (vsmutok) [inline app/customer.py]:
```suggestion
            print(
                f"{self.name}'s trip to the "
                f"{shop.name} costs {round(total_cost, 2)}"
            )
```

File: app/customer.py [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+            print(f"{self.name} rides to {cheapest_shop.name}\n")
+
+            self.make_purchase(cheapest_shop)
+            date = (datetime(
+                2021, 4, 1, 12, 33, 41
+            ).strftime("%m/%d/%Y %H:%M:%S"))
+            print(f"Date: {date}\n"
+                  f"Thanks, {self.name}, for your purchase!\n"
+                  f"You have bought:"
+                  )
```
Comment (vsmutok) [inline app/customer.py]:
```suggestion
            date = datetime(
                2021, 4, 1, 12, 33, 41
            ).strftime("%m/%d/%Y %H:%M:%S")
            print(
                f"Date: {date}\n"
                f"Thanks, {self.name}, for your purchase!\n"
                f"You have bought:"
            )
```

File: app/customer.py [hunk 13]
```diff
@@ -0,0 +1,78 @@
...
+                    int(total_price) if total_price == int(total_price)
+                    else float(total_price)
+                )
+                print(f"{self.product_cart[product]} "
+                      f"{product}s for {total_price} dollars")
+            print(f"Total cost is "
+                  f"{self.make_purchase(cheapest_shop)} dollars\n"
+                  f"See you again!\n\n"
+                  f"{self.name} rides home"
+                  )
```
Comment (vsmutok) [inline app/customer.py]:
```suggestion
                print(
                    f"{self.product_cart[product]} "
                    f"{product}s for {total_price} dollars"
                )
            print(
                f"Total cost is "
                f"{self.make_purchase(cheapest_shop)} dollars\n"
                f"See you again!\n\n"
                f"{self.name} rides home"
            )
```

File: app/customer.py [hunk 14]
```diff
@@ -0,0 +1,78 @@
...
+            self.money -= min_cost
+            print(f"{self.name} now has {round(self.money, 2)} dollars\n")
+
+        else:
+            print(f"{self.name} doesn't have enough money "
+                  f"to make a purchase in any shop")
```
Comment (vsmutok) [inline app/customer.py]:
```suggestion
            print(
                f"{self.name} doesn't have enough money "
                f"to make a purchase in any shop"
            )
```

File: app/main.py [hunk 15]
```diff
@@ -1,3 +1,37 @@
...
+    customers_data = config.get("customers")
+    customers = [
+        Customer(
+            name=customer_data.get("name"),
+            product_cart=customer_data.get("product_cart"),
+            location=customer_data.get("location"),
+            money=customer_data.get("money"),
+            car=Car(**customer_data.get("car")
+                    )
+        ) for customer_data in customers_data
```
Comment (vsmutok) [inline app/main.py]:
```suggestion
            car=Car(**customer_data.get("car")),
        )
        for customer_data in customers_data
```

File: app/main.py [hunk 16]
```diff
@@ -1,3 +1,37 @@
...
+from app.shop import Shop
+from app.customer import Customer
+
+
+def shop_trip() -> None:
+    try:
+        with open("app/config.json", "r") as file:
+            config = json.load(file)
+    except FileNotFoundError:
+        raise
```
Comment (vsmutok) [inline app/main.py]:
```suggestion
        raise Exception("some clear message")
```

**Review by vsmutok — CHANGES_REQUESTED** [review]

**Review by vsmutok — APPROVED** [review]
Well Done!

**Pull request closed without merging**